tests/test_lignin_kmc_parts.py

```diff
@@ -0,0 +1,874 @@
+    initial_state = create_initial_state(initial_events, initial_monomers)
+    initial_events.append(Event(GROW, [], rate=1e4))
+    result = run_kmc(DEF_RXN_RATES, initial_state, sorted(initial_events), n_max=max_monos, t_max=2, random_seed=seed)
+    return result
+
+
+# Tests #
+
+class TestCalcRates(unittest.TestCase):
+    """
+    Tests calculation of rate coefficients by the Eyring equation.
+    """
+    def testCalcRatesFromKcalMol(self):
+        rxn_rates = calc_rates(DEF_TEMP, ea_kcal_mol_dict=DEF_E_BARRIER_KCAL_MOL)
+        self.assertTrue(len(rxn_rates) == len(DEF_RXN_RATES))
+        rxn_type, substrate, substrate_type = None, None, None  # to make IDE happy
+        try:
+            for rxn_type in DEF_RXN_RATES:
+                for substrate in DEF_RXN_RATES[rxn_type]:
+                    for substrate_type in DEF_RXN_RATES[rxn_type][substrate]:
+                        self.assertAlmostEqual(DEF_RXN_RATES[rxn_type][substrate][substrate_type],
+                                               rxn_rates[rxn_type][substrate][substrate_type])
+        except (TypeError, IndexError) as e:
+            print(f'{e}\nError when looking at rxn_type: {rxn_type} substrate: {substrate}    '
+                  f'substrate_type:    {substrate_type}')
+
+
+class TestMonomers(unittest.TestCase):
+    def testCreateConiferyl(self):
+        mon = Monomer(0, 0)  # Makes a guaiacol type monomer with ID = 0
+        self.assertTrue(mon.open == {8, 4, 5})
+        self.assertTrue(str(mon) == '0: coniferyl alcohol is connected to {0} and active at position 0')
+
+    def testCreateSyringol(self):
+        mon = Monomer(1, 2)  # Makes a syringol type monomer with ID = 2
+        self.assertTrue(mon.open == {4, 8})
+        self.assertTrue(mon.connectedTo == {2})
+        self.assertTrue(str(mon) == '2: sinapyl alcohol is connected to {2} and active at position 0')
+        self.assertTrue(repr(mon) == '2: sinapyl alcohol \n')
+
+    def testUnknownUnit(self):
+        try:
+            mon = Monomer(3, 2)
+            # type type 3 is not currently implemented
+            self.assertFalse(mon)  # should not be reached
+        except InvalidDataError as e:
+            self.assertTrue("only the following" in e.args[0])
+
+    def testHash(self):
+        mon1 = Monomer(1, 5)
+        mon2 = Monomer(1, 5)
+        check_set = {mon1, mon2}
+        self.assertTrue(len(check_set) == 1)
+
+
+class TestEvent(unittest.TestCase):
+    def testIDRepr(self):
+        rxn = OX
+        # noinspection PyTypeChecker
+        event1 = Event(rxn, [2], DEF_RXN_RATES[rxn][0][MONOMER])
+        self.assertTrue(str(event1) == "Performing oxidation on index 2")
+
+    def testIDReprBond(self):
+        rxn = BO4
+        # noinspection PyTypeChecker
+        event1 = Event(rxn, [1, 2], DEF_RXN_RATES[rxn][(0, 1)][MON_OLI], (4, 5))
+        good_str = "Forming bo4 bond between indices [1, 2] (adjacency_matrix update (4, 5))"
```

Comment:
Was there a reason you wanted to make an event that was named BO4 but that made a 5O4 bond? I recognize that during my initial implementation, I used this repetitive `bond` parameter, but it seems like we should be able to eliminate that (maybe in a future branch). Seems like the most concise way to represent events would be to just rely on a key (bond name), reaction rate, and pair of monomers that it would affect. (We would probably also want flags for having a reverse bond, and monomer ratios for growth events)

Reply:
I think I took this from the code as I saw it. We can fix in another iteration, if you wish.